File: Stack/Java/Asteroid Collision.java
// Problem link: https://leetcode.com/problems/asteroid-collision/description/
// Time: O(n)
// Space: O(n)
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> stack = new Stack<>();

        for(int idx = 0; idx < asteroids.length; idx++) {
            while(!stack.isEmpty() && asteroids[idx] < 0 && stack.peek() > 0) {
                if(asteroids[idx] + stack.peek() < 0) {
                    stack.pop();
                }
                else if(asteroids[idx] + stack.peek() > 0) {
                    asteroids[idx] = 0;
                }
                else {
                    stack.pop();
                    asteroids[idx] = 0;
                }
            }

            if(asteroids[idx] != 0)
                stack.push(asteroids[idx]);
        }

        int len = stack.size();
        int[] result = new int[len];

        for(int idx = len - 1; idx >= 0; idx--)
            result[idx] = stack.pop();
        
        return result;
    }
}

File: Stack/Python/Asteroid Collision.py
# Problem link: https://leetcode.com/problems/asteroid-collision/description/
# Time: O(n)
# Space: O(n)
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []

        for ast in asteroids:
            while ast < 0 and stack and stack[-1] > 0:
                if ast + stack[-1] > 0:
                    ast = 0
                elif ast + stack[-1] < 0:
                    stack.pop()
                else:
                    ast = 0
                    stack.pop()
            if ast:
                stack.append(ast)
        
        return stack
